pallets/utxo/src/rewards.rs

```diff
@@ -0,0 +1,207 @@
+            f.try_into().unwrap_or(u8::MAX)
+        })
+}
+
+
+/// Returns the newly reduced reward amount for a Block Author.
+/// How much a reward is reduced, will be based on the config's`RewardReductionFraction`.
+fn get_block_author_reward<T: Config>(block_number: T::BlockNumber) -> Value {
+    let reward_amount = T::InitialReward::get();
+    let reduction_period: T::BlockNumber = T::RewardReductionPeriod::get();
+    let reduction_fraction = T::RewardReductionFraction::get().deconstruct();
+
+    // no computation needed; current block number is still within the reduction period.
+    if block_number < reduction_period {
+        return reward_amount;
+    }
+
+    // The maximum period before approaching to 100% reduction
+    let max_period = (100u8 / reduction_fraction) - 1;
+
+    // current period of the given block number
+    let current_period = get_block_period::<T>(block_number);
+
+    // when current_period has reached or exceeded the maximum period of reduction,
+    // return the least amount to reward.
+    if current_period > max_period {
+        // TODO: as a note, this is only testnet specific to reward at least 1
+        return 1;
+    }
+
+    // compute for the updated reduction % , given the current period.
+    let updated_reduction_fraction = Percent::from_percent(reduction_fraction * current_period );
+
+    reward_amount - updated_reduction_fraction.mul_ceil(reward_amount)
+}
+
+fn insert_to_utxo_store<T:Config>(block_number:T::BlockNumber, block_author:H256, reward:Value) {
+    let utxo = TransactionOutput::new_pubkey(reward, block_author.clone());
+
+    let hash = {
+        let b_num = block_number.saturated_into::<u64>();
+        BlakeTwo256::hash_of(&(&utxo, b_num, "author_reward"))
+    };
+
+    if !<UtxoStore<T>>::contains_key(hash) {
+        <UtxoStore<T>>::insert(hash, utxo.clone());
+
+        <Pallet<T>>::deposit_event(Event::<T>::BlockAuthorRewarded(utxo));
+    }
+}
+
+/// Rewards the block author with a utxo of value based on the `BlockAuthorRewardAmount`
+/// and the transaction fees.
+pub(crate) fn reward_block_author<T: Config>(block_number: T::BlockNumber) {
+
+    // As written on the definition of Take:
+    // Take a value from storage, removing it afterwards.
+    // This is taking a value of the RewardTotal storage, freeing it up.
+    let transaction_fees = <RewardTotal<T>>::take();
+
+    if let Some(reward_amount) = get_block_author_reward::<T>(block_number).checked_add(transaction_fees) {
+        // As written on the definition of Take:
+        // Take a value from storage, removing it afterwards.
+        // This is taking a value of the BlockAuthor storage, freeing it up.
+        match <BlockAuthor<T>>::take() {
+            None => {
+                log::warn!("no block author found for block number {:?}", block_number);
+                // carry over the fees to the next block rewarding.
+                <RewardTotal<T>>::put(reward_amount);
```

Comment:
The reason why this failure is bad and should never happen, is because we're moving the fees from one block to another... very weird way of doing things.